Document how to add features to the residential model (#39)
ccao-data/wiki: How-To/Create-a-New-Residential-Feature.md

-   Choose the output location based on your transformation strategy. If the file is already clean, and a SQL query can be easily implemented on it, the script should be stored in `ccao-data-warehouse-us-east-1` . If the file needs additional cleaning, place it in `ccao-data-raw-us-east-1`. Use a modification of the following script to upload the data to S3 in python. Note how it joins the aforementioned AWS_S3_RAW_BUCKET, as well as identifying the correct sub-bucket, in this case `housing`, before creating the unique folder `dci`.

```         
def upload_to_s3(file_content, bucket, key_prefix, file_name):
    try:
        s3 = boto3.resource("s3")
        object_key = f"{key_prefix}/{file_name}"
        s3.Bucket(bucket).put_object(Key=object_key, Body=file_content)
        return f"File uploaded to {bucket}/{object_key}"
        file_name = os.path.join("housing", "dci", "dci.csv") 
        upload_df_to_s3(df, AWS_S3_RAW_BUCKET, file_name)
```

> -   If you export the data to `ccao-data-warehouse-us-east-1` for direct use in an SQL / DBT model, make sure that the data is written as a parquet file. If it is written to `ccao-data-raw-us-east-1`, it can be either parquet or csv.

Author: I think that the issue is more how we partition and run macros in certain structures where it is good to have files standardized in certain formats at this stage in the process.

```         
{{config(          
materialized='table',         
partitioned_by=['year'],          
bucketed_by=['pin10'],          
bucket_count=1)}
```

If the output does not have the partition column as the final column, you will receive the following error:

```         
HIVE_COLUMN_ORDER_MISMATCH: Partition keys must be the last columns in the table and in the same order as the table properties: 
```

-   Build out the transformation. To ensure consistency, use one of the macros whenever possible, each of which utilizes a specific coding structure. For example, the `dist_to_nearest_geometry.sql` requires a value for `pin`, `year`, `x_3435`, and `y_3435`. These also need to be structured in accordance with specific data types. In this case, if geographies are classified as a string, you will run into the following error.

```         
Runtime Error in model spatial.spatial_stadium (models/spatial/spatial.spatial_stadium.sql) line 9:5: Unknown type: STRING
```

To remedy this, you can recast the geometry value as a binary points.

```         
ST_ASBINARY(ST_POINT(stadium.lon, stadium.lat)) AS geometry
```

> -   Once the macro / sql query is correct, run it in your terminal to construct the output in your development database of `AWS Athena`. The output view is the file which will be referenced during `Step 4`. A sample build is presented below.

Reviewer: **[Nitpick, non-blocking]** Some slightly clearer language:

```suggestion
-   Once your model is defined correctly, run `dbt build` in your terminal to construct the output in your development database:
```

```         
dbt build --select proximity.dist_pin_to_stadium
```

> -   Complete the documentation by configuring it as a [source](https://docs.getdbt.com/docs/build/sources) in the folder's dbt `schema.yml` file, and update the folder's `docs.md` and `columns.md` files.

Reviewer: **[Nitpick, non-blocking]** I believe the table created by this step should be configured as a model, not a source -- the raw data that is used in the transformation (either from `ccao-data-raw-us-east-1` or `ccao-data-warehouse-us-east-1`) but whose extract/transform script is defined elsewhere should be a source, but any transformations defined in dbt should be models.

Reviewer: @Damonamajor Looks like this suggestion still needs to be incorporated!

------------------------------------------------------------------------

## Step 4: Incorporate the Transformed Data into the [Final Model View](https://github.com/ccao-data/data-architecture/tree/master/dbt/models/model).

-   Now that the relevant view(s) have been created, they need to be incorporated in the model view.

> -   Identify if the new feature is relevant in the condo and/or the residential model. Then, in `dbt/models/model`, update the relevant `vw_pin_condo_input`, `vw_res_card_input`, and `vw_shared_input` view. Within the `model` directory, update the definitions in the appropriate dbt model.

Reviewer: **[Nitpick, non-blocking]** The last sentence here feels duplicative of the one that precedes it, but maybe I'm missing something:

```suggestion
-   Identify if the new feature is relevant in the condo and/or the residential model. Then, in `dbt/models/model`, update the relevant `vw_pin_condo_input`, `vw_res_card_input`, or `vw_shared_input` view.
```

> -   If the attribute is spatial, you should also update `crosswalk_year_fill.sql`, `proximity.vw_pin10_proximity_fill.sql`, and `location.vw_pin10_location_fill.sql`. Processes for forward and backward filling are largely conceptual. If we think that the data is time invariant, we can forward fill the data. If we believe that the data has historically been the same, we can backwards fill the data.

Reviewer: **[Question, non-blocking]** What's the difference between being "time invariant" and "historically [being] the same"? I wonder if this is actually the same phenomenon such that we don't need to draw a distinction between forward and backward filling? I'm also not totally sure what it means for them to be "largely conceptual". If you agree, clearer might be:

```suggestion
-   If the attribute is spatial, you should also update `crosswalk_year_fill.sql`, `proximity.vw_pin10_proximity_fill.sql`, and `location.vw_pin10_location_fill.sql`. Make sure to fill in missing years of data: If certain years are missing from the data but we think that it did not change during that period, we can forward- or backward-fill it.
```

------------------------------------------------------------------------

## Step 5: Update the [Model Pipeline](https://github.com/ccao-data/model-res-avm/tree/master) to use the New Versions of the Model Input Views

> -   Create a new issue / pull request in the `model-res-avm` repository.

Reviewer: **[Question, non-blocking]** What does creating a new issue do for us here? I wonder if we can assume that the reader will already be assigned to an issue if they're reading these docs? And perhaps "branch" is clearer than "pull request", since no code changes have been made to `model-res-avm` at this point?

```suggestion
-   Create a new branch in the `model-res-avm` repository.
```

-   Add the new feature to the `predictor` section of `model/models/params.yaml`.

-   If you want to test the impact, you can run the [model ingest steps](https://github.com/ccao-data/model-res-avm?tab=readme-ov-file#usage) with an updated versin of `model.vw_card_res_input`. Make sure that you unfreeze the desired `stage(s)` with `dvc unfreeze {stage}`.

-   To test the impact, you can run the full model with the command

```         
dvc repro -f
```

> -   If you want this information to be shared, you can push the updated model data to DVC using the following code. If you do this, make sure that the param for run_type is `test`.

Reviewer: **[Nitpick, non-blocking]** Let's be more specific about what this means:

```suggestion
-   If you want others to be able to reproduce your work, you can push the updated model data to DVC using the following commands. If you do this, make sure to set the `run_type to `test` in `params.yaml`.
```

```         
dvc commit
dvc push 
```

------------------------------------------------------------------------

## Troubleshooting

> -   Some staff types do not have permission to write to AWS Buckets. If your export of data to `aws` returns `Access Denied`, ask senior staff to run the query for you. This is most likely to occur during step 1.

Reviewer: **[Nitpick, non-blocking]** Some minor formatting and clarity issues:

```suggestion
-   Some staff members do not have permission to write to AWS buckets. If you run into an `Access Denied` error at any point, ask senior staff to run your code for you. This is most likely to occur during step 1.
```

> -   Make sure that you are connected to aws by running `aws-mfa` in the terminal and typing in the correct credentials.

Reviewer: **[Nitpick, non-blocking]** `aws-mfa` has a purpose that's slightly more specific than "connection":

```suggestion
-   Make sure that you have authenticated to AWS using MFA by running `aws-mfa` in the terminal and typing in the correct credentials.
```

> -   Whenever an active pull request is open, committing your file build all modified views. If a file is not constructed, you will receive an error message. In the following, the sql query is identifying that cyf, a reference to `crosswalk_year_fill.sql` not being properly built, a prerequisite for building `vw_pin10_proximity_fill.sql`.

Reviewer: **[Nitpick, non-blocking]** Minor typos and clarity:

```suggestion
-   Whenever you commit to an open pull request against `data-architecture`, the `build-and-test-dbt` workflow will build all tables and views you have modified. If a table or view that the modified model depends on has not been built yet, you will receive an error message. In the following, the SQL query is identifying that `cyf`, a reference to the `proximity.crosswalk_year_fill` model, has not been built but is a prerequisite for building `vw_pin10_proximity_fill.sql`.
```

```         
line 176:9: Column 'cyf.nearest_stadium_data_year' cannot be resolved
```

> -   Relevant docs also need to be properly constructed. If a file cannot be properly identified, you will receive an error like this.

Reviewer: **[Nitpick, non-blocking]** Clarity:

```suggestion
-   Relevant docs also need to be properly constructed. If a [docs block](https://docs.getdbt.com/docs/build/documentation#using-docs-blocks) cannot be properly identified, you will receive an error like this.
```

-   Choose the output location based on your transformation strategy. If the file is already clean, and a SQL query can be easily implemented on it, the script should be stored in `ccao-data-warehouse-us-east-1`. If the file needs additional cleaning, place it in `ccao-data-raw-us-east-1`. Use a modification of the following script to upload the data to S3 in python. Note how it joins the aforementioned `AWS_S3_RAW_BUCKET`, as well as identifying the correct prefix, in this case `/housing`, before creating the unique folder `dci`.

```         
def upload_to_s3(file_content, bucket, key_prefix, file_name):
    s3 = boto3.resource("s3")
    object_key = f"{key_prefix}/{file_name}"
    s3.Bucket(bucket).put_object(Key=object_key, Body=file_content)
    return f"File uploaded to {bucket}/{object_key}"

upload_to_s3(df, AWS_S3_RAW_BUCKET, "housing/dci", "dci.csv")

> ```

Reviewer: **[Nitpick, non-blocking]** Same as above, this code block could benefit from a `python` language identifier. (I'd add a suggestion for this change, but they don't play well with code blocks, since both suggestions and code blocks use triple-backticks to mark the start and end of blocks.)

## Step 1: Extract and Load the Raw Data

### Option A: Use a Python or R Script in the [etl Directory](https://github.com/ccao-data/data-architecture/tree/master/etl)

Data extraction scripts are typically created in R or Python in the `etl/scripts-ccao-data-raw-us-east-1/` folder. In the circumstance that the file requires no cleaning afterwards, they can be placed directly in the `scripts-ccao-data-warehouse-us-east-1` folder. Data sources often contain information which is relevant for institutional knowledge, but are not useful for modeling. Keeping data in its raw form provides redundancy in case there are complications from the data transformation or the data source changes over time.

-   Identify which folder the script should go in. Although folders are often self-explanatory, the location may switch throughout the feature creation process. For example, `spatial/spatial-environment-secondary_road.R` is created as a `spatial` feature, but during the transformation step (`proximity.dist_pin_to_secondary_roads.sql)`, it shifts to the `proximity` folder since the metric (distance) is in relation to PINs.

> -   Set the `AWS_S3_RAW_BUCKET` environment variable to an S3 bucket where output files will be stored:

Reviewer: **[Question, blocking]** This seems like a duplicate of line 30, is that right? Should we say something to introduce the imports here instead?

```         
import os
import boto3
from dotenv import load_dotenv
```

```         
library(arrow)
library(aws.s3)

> ```

Reviewer: **[Suggestion, non-blocking]** We could add [language identifiers](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) to these code blocks to instruct GitHub to add syntax highlighting.

    ```python
    import os
    import boto3
    from dotenv import load_dotenv
    ```

    ```r
    library(arrow)
    library(aws.s3)
    ```

```         
def upload_to_s3(file_content, bucket, key_prefix, file_name):
    s3 = boto3.resource("s3")
    object_key = f"{key_prefix}/{file_name}"
    s3.Bucket(bucket).put_object(Key=object_key, Body=file_content)
    return f"File uploaded to {bucket}/{object_key}"

upload_to_s3(df, AWS_S3_RAW_BUCKET, "housing/dci", "dci.csv")
```

-   If you export the data to `ccao-data-warehouse-us-east-1` for direct use in a dbt model, make sure that the data is written as a parquet file. If it is written to `ccao-data-raw-us-east-1`, it can be either parquet or csv.
-   Now that the file has been built, crawl the output with [Glue](https://us-east-1.console.aws.amazon.com/glue/home?region=us-east-1#/v2/data-catalog/crawlers) by navigating to the crawler that is configured to crawl your bucket and then selecting `Run Crawler` in the upper right.

> ## Option B: Use a Seed to Create the Data in \`[dbtseeds](https://github.com/ccao-data/data-architecture/tree/master/dbt/seeds)\`

Reviewer: **[Nitpick, non-blocking]** I think we are possibly missing a slash here, right? We also want to use a third-level header if we want to match the `Option A` section above.

```suggestion
### Option B: Use a Seed to Create the Data in \`[dbt/seeds](https://github.com/ccao-data/data-architecture/tree/master/dbt/seeds)\`
```

Author: These "\`" were not rendering well for me. Happy to add them back in if there was a purpose, but I took them out.

Reviewer: Good call, I think the text looks better without the escaped backticks.

------------------------------------------------------------------------

## Step 4: Incorporate the Transformed Data into the [Final Model View](https://github.com/ccao-data/data-architecture/tree/master/dbt/models/model).

-   Now that the relevant view(s) have been created, they need to be incorporated in the model view.

-   Identify if the new feature is relevant in the condo and/or the residential model. Then, in `dbt/models/model`, update the relevant `vw_pin_condo_input`, `vw_res_card_input`, or `vw_shared_input` view.

-   If the attribute is spatial, you should also update `crosswalk_year_fill.sql`, `proximity.vw_pin10_proximity_fill.sql`, and `location.vw_pin10_location_fill.sql`. Make sure to fill in missing years of data: If certain years are missing from the data but we think that it did not change during that period, we can forward- or backward-fill it.

------------------------------------------------------------------------

## Step 5: Update the [Model Pipeline](https://github.com/ccao-data/model-res-avm/tree/master) to Use the New Versions of the Model Input Views

-   Create a new branch in the `model-res-avm` repository.

> -   Add the new feature to the `predictor` section of `model/models/params.yaml`.

Reviewer: **[Nitpick, blocking]** I think `params.yaml` lives at the top level of the model repo, right?

```suggestion
-   Add the new feature to the `predictor` section of `params.yaml`.
```

-   If you want others to be able to reproduce your work, you can push the updated model data to DVC using the following commands. If you do this, make sure to set the `run_type` to `test` in `params.yaml`.

-   To test the impact, you can run the full model with the command

```         
dvc repro -f
```

-   If you want others to be able to reproduce your work, you can push the updated model data to DVC using the following commands. If you do this, make sure to set the `run_type to `test` in `params.yaml`.

```         
dvc commit
dvc push 
```

------------------------------------------------------------------------

## Troubleshooting

> -   Some staff members do not have permission to write to AWS buckets. If you run into an `Access Denied` error at any point, ask senior staff to run your code for you. This is most likely to occur during step 1.

Reviewer: **[Suggestion, non-blocking]** Based on Michael's recent experience, this will probably also happen during `dvc push` in step 5, right?

    * Option 2: Add a Python/R script to `etl/scripts-ccao-data-warehouse-us-east-1/`
        * This requires that the raw data you pushed in step 1 live in the `ccao-data-raw-us-east-1` bucket
        * This is a good option if your transformation is complex enough to require a Python or R script using third-party packages beyond what is built-in to Athena PySpark
        * If you choose this option, you'll need to upload your file as a parquet, configure your output table as a [source](https://docs.getdbt.com/docs/build/sources) in a dbt `schema.yml` file, and set up a Glue crawler for it
3. Incorporate the transformed data into the model input views
    * The views in question include `vw_pin_condo_input`, `vw_res_card_input`, and `vw_shared_input`; choose the correct view depending on whether the feature should be available to the condo model, the res model, or both, and update the view definition in the appropriate dbt model to join to the transformation table you created in step 2

>     * If the attribute is spatial, you should also update `crosswalk_year_fill.sql` and `proximity.vw_pin10_proximity_fill.sql` / `location.vw_pin10_location_fill.sql`

Author: Added line

## Collective Steps

### Step 1 - Run Crawler for Ingest into AWS Athena.
Once final data has been uploaded to the AWS S3, it needs to be transfered to [AWS Athena](https://us-east-1.console.aws.amazon.com/athena/home?region=us-east-1#/query-editor) via the [crawler](https://us-east-1.console.aws.amazon.com/glue/home?region=us-east-1#/v2/data-catalog/crawlers). Find the correct bucket, and then click 'Run Crawler' in the upper right. 

### Step 2 - Include in the Model Pipeline

To include the feature in the model pipeline, the sql queries vw_pin_condo_input, vw_res_card_input, and vw_shared_input located in [data_architecture/dbt/models/model/](https://github.com/ccao-data/data-architecture/tree/master/dbt/models/model) need to be modified.

### Step 3 - Update Documentation

CCAO data is largely mantained through DBT. To ensure data integrity, documentation needs to be updated alongside the new feature. These can be found in data_architecture/dbt/models/model/schema.yml, dbt/models/shared_columns.md, dbt/models/.../docs.md, and dbt/models/.../schema.yml.

## Final Step - An Optional Horrah

> Test the impact of the feature by running the model via the following [guide](https://github.com/ccao-data/model-res-avm/blob/master/README.md). 

Reviewer: **[Suggestion, blocking]** I think the "Structure A / Structure B / Collective Steps" organizational outline could be a bit clearer, and I think we're missing a couple of steps toward the end that tie the input views to the model itself. Here's how I would outline the process:

1. Extract and load raw data
    * Do this by adding a raw data extraction script to `etl/scripts-ccao-data-raw-us-east-1/` using Python or R
    * Choose the output location based on the transformation strategy in step 2: The script should push data to `ccao-data-warehouse-us-east-1` if transforming the data using a dbt model, or `ccao-data-raw-us-east-1` if transforming the data using a Python/R script
2. Transform the raw data
    * Option 1: [Add a model to the dbt DAG](https://github.com/ccao-data/data-architecture/blob/master/dbt/README.md#-how-to-add-a-new-model) to transform the data
        * This requires that the raw data you pushed in step 1 live in the `ccao-data-warehouse-us-east-1` bucket, be crawled by Glue, and be configured as a [source](https://docs.getdbt.com/docs/build/sources) in a dbt `schema.yml` file
        * This is a good option if your transformation is simple enough to be written as a SQL query, or as a Python script using Athena PySpark's [built-in third-party packages](https://docs.aws.amazon.com/athena/latest/ug/notebooks-spark-preinstalled-python-libraries.html)
    * Option 2: Add a Python/R script to `etl/scripts-ccao-data-warehouse-us-east-1/`
        * This requires that the raw data you pushed in step 1 live in the `ccao-data-raw-us-east-1` bucket
        * This is a good option if your transformation is complex enough to require a Python or R script using third-party packages beyond what is built-in to Athena PySpark
        * If you choose this option, you'll need to configure your output table as a [source](https://docs.getdbt.com/docs/build/sources) in a dbt `schema.yml` file and set up a Glue crawler for it
3. Incorporate the transformed data into the model input views
    * The views in question include `vw_pin_condo_input`, `vw_res_card_input`, and `vw_shared_input`; choose the correct view depending on whether the feature should be available to the condo model, the res model, or both, and update the view definition in the appropriate dbt model to join to the transformation table you created in step 2
4. Update each model to use the new versions of the model input views
    * For each model:
        1. Rerun the model ingest steps to pull from the new versions of the input views
        2. Push the updated model input data to DVC
        3. Update the `predictor` attribute in `params.yml` to add the new columns as predictors
